numpy/core/tests/test_multiarray.py

```diff
@@ -2041,6 +2041,13 @@ def __numpy_ufunc__(self, ufunc, method, pos, inputs, **kwargs):
         assert_raises(TypeError, np.dot, b, c)
         assert_raises(TypeError, c.dot, b)
 
+    def test_dot_type_mismatch(self):
+        c = 1.
+        A = np.array((1,1), dtype='i,i')
+
+        assert_raises(ValueError, np.dot, c, A)
+        assert_raises(TypeError, np.dot, A, c)
```

Comment:
So, I managed to get this built so I can debug with `python3-dbg` interactively and discovered this line will cause the strange exception. However, the line before does not cause the exception.